_drafts/firmware-watchdog-best-practices.md

```diff
@@ -0,0 +1,796 @@
+
+```
+(gdb) CTRL-C
+[... wait 7 seconds ...]
+(gdb) continue
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+0x00002bbc in watchdog_fault_handler_c (frame=0x20002a78) at interrupt/example/watchdog-example/software_watchdog.c:81
+81    HALT_IF_DEBUGGING();
+```
+
+This is a situation where we can leverage GDBs Python API! (For a deeper dive into the GDB Python API check out [this post]({% post_url 2019-07-02-automate-debugging-with-gdb-python-api %})).
+
+Specifically, we can make use of the Events API[^9], to pause
+the peripheral timer when the system halts and resume it when the system continues.
+
+#### Example Code to Pause Software Watchdog While Debugging
+
+```python
+# gdb_resume_handlers.py
+try:
+    import gdb
+except ImportError:
+    raise Exception("This script can only be run within gdb!")
+
+import struct
+
+
+class Nrf52Timer0:
+    TIMER0_TASKS_START_ADDR = 0x40008000
+    TIMER0_TASKS_STOP_ADDR = 0x40008004
+
+    @staticmethod
+    def _trigger_task(addr):
+        gdb.selected_inferior().write_memory(addr, struct.pack("<I", 1))
+
+    @classmethod
+    def start(cls):
+        print("Resuming Software Watchdog")
+        cls._trigger_task(cls.TIMER0_TASKS_START_ADDR)
+
+    @classmethod
+    def stop(cls):
+        print("Pausing Software Watchdog")
+        cls._trigger_task(cls.TIMER0_TASKS_STOP_ADDR)
+
+
+def nrf52_debug_stop_handler(event):
+    Nrf52Timer0.stop()
+
+
+def nrf52_debug_start_handler(event):
+    Nrf52Timer0.start()
+
+
+gdb.events.stop.connect(nrf52_debug_stop_handler)
+gdb.events.cont.connect(nrf52_debug_start_handler)
+```
+
+Now when we invoke GDB we can add the argument `--ex="source gdb_resume_handlers.py"` and the timer will be automatically paused by GDB for us while the debugger is halted.
+
+```
+arm-none-eabi-gdb-py --eval-command="target remote localhost:2331"  --ex="mon reset" --ex="load" --ex="mon reset" --ex="source gdb_resume_handlers.py" --se=build/nrf52.elf
+Program received signal SIGTRAP, Trace/breakpoint trap.
+main () at interrupt/example/watchdog-example/main.c:134
+134   }
+Pausing Software Watchdog
+(gdb) c
+Continuing.
+Resuming Software Watchdog
+```
+
+## Debugging Hangs Inside Interrupts
+
+Let's see how the system fairs if a hang occurs within an ISR:
+
+```
+(gdb) set g_watchdog_hang_config=4
+(gdb) c
+Continuing.
+Resuming Software Watchdog
+(gdb) p/a *frame
+$2 = {
+  r0 = 0x3 <g_pfnVectors+3>,
+  r1 = 0x200001bc <ucHeap+84>,
+  r2 = 0xe000e100,
+  r3 = 0x1 <g_pfnVectors+1>,
+  r12 = 0x200003e0 <ucHeap+632>,
+  lr = 0xfffffffd,
+  return_address = 0xd0 <ExternalInt0_Handler>,
+  xpsr = 0x21000010
+}
+```
+
+The return address points to an interrupt handler, `ExternalInt0_Handler`. Inspecting the code it does indeed hang:
+
+```c
+void ExternalInt0_Handler(void) {
+  while (1) {}
+}
+```
+
+### ARM Cortex-M Design Suggestions
+
```

Comment:
Add a sentence here, like: "Over the years, I've developed a set of best practices for implementing watchdogs on cortex-M MCUs".